### PR TITLE
Feat/tr 3118/async result item transmission

### PR DESCRIPTION
**Pull request by poyuki**
# [TR-3118](https://oat-sa.atlassian.net/browse/TR-3118)

## Goal
To move storing `result variable` in asynchronous flow  

## Changelog
- feat: add functionality which allow providing result transmission as synchronous and asynchronous event
- doc: add new functionality configuration description into README.md 

## How to test
1. Install current branch on `CG`  instance
2. Run `php tao/scripts/taoUpdate.php -vvv` of `php tao/scripts/taoSetup.php setup.json -vvv`
3.  Initialize default task queue
4. Update DI conf `config/taoQtiTest/ResultTransmissionEventHandler.conf.php` by next code 
```PHP
<?php
/**
 * Default config header created during install
 */

return new oat\taoQtiTest\models\classes\evnetHandler\ResultTransmissionEventHandler\AsynchronousResultTransmissionEventHandler();
```
Results of execution will be provided in result page and next messages can be found in logs:
- Result test variable successfully transmitted for deliver execution - %s
- Results item variable successfully transmitted for deliver execution - %s

### CODE REVIEW

File: migrations/Version202112231326462260_taoQtiTest.php [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+use oat\oatbox\event\EventManager;
+use oat\oatbox\reporting\Report;
+use oat\tao\scripts\tools\migrations\AbstractMigration;
+use oat\taoQtiTest\models\classes\event\ResultTestVariablesTransmissionEvent;
+use oat\taoQtiTest\models\classes\evnetHandler\ResultTransmissionEventHandler\Api\ResultTransmissionEventHandlerInterface;
+use oat\taoQtiTest\models\event\ResultItemVariablesTransmissionEvent;
+
+/**
+ * Auto-generated Migration: Please modify to your needs!
+ */
```
Comment (boajer) [inline migrations/Version202112231326462260_taoQtiTest.php:17]:
```suggestion
```

File: migrations/Version202112231326462260_taoQtiTest.php [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    {
+        $eventManager->attach(
+            ResultItemVariablesTransmissionEvent::class,
+            [ResultTransmissionEventHandlerInterface::SERVICE_ID, 'transmitResultItemVariable']
+        );
+
+        $this->addReport(
+            Report::createSuccess(
+                sprintf(
+                    'Unsubscribed %s::%s to %s event',
```
Comment (boajer) [inline migrations/Version202112231326462260_taoQtiTest.php:51]: Shouldn't be like that?
```suggestion
                    'Subscribed %s::%s to %s event',
```

File: migrations/Version202112231326462260_taoQtiTest.php [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    {
+        $eventManager->attach(
+            ResultTestVariablesTransmissionEvent::class,
+            [ResultTransmissionEventHandlerInterface::SERVICE_ID, 'transmitResultTestVariable']
+        );
+
+        $this->addReport(
+            Report::createSuccess(
+                sprintf(
+                    'Unsubscribed %s::%s to %s event',
```
Comment (boajer) [inline migrations/Version202112231326462260_taoQtiTest.php:88]: Shouldn't be like that?

```suggestion
                    'Subscribed %s::%s to %s event',
```

File: models/classes/event/ResultItemVariablesTransmissionEvent.php [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+declare(strict_types=1);
+
+namespace oat\taoQtiTest\models\event;
+
+use oat\oatbox\event\Event;
+
+class ResultItemVariablesTransmissionEvent implements Event
+{
+    /** @var array */
+    private $variables;
```
Comment (boajer) [inline models/classes/event/ResultItemVariablesTransmissionEvent.php:30]:
```suggestion
    private $variables = [];
```

File: models/classes/event/ResultTestVariablesTransmissionEvent.php [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+namespace oat\taoQtiTest\models\classes\event;
+
+use oat\oatbox\event\Event;
+
+class ResultTestVariablesTransmissionEvent implements Event
+{
+    /** @var string */
+    private $deliveryExecutionId;
+    /** @var array */
+    private $variables;
```
Comment (boajer) [inline models/classes/event/ResultTestVariablesTransmissionEvent.php:32]:
```suggestion
    private $variables = [];
```

File: models/classes/evnetHandler/ResultTransmissionEventHandler/AsynchronousResultTransmissionEventHandler.php [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    private function getQueueDispatcher(): QueueDispatcher
+    {
+        return ServiceManager::getServiceManager()->get(QueueDispatcher::SERVICE_ID);
+    }
+
+    private function packVariables(array $variables): array
+    {
+        return array_map(static function ($record) {
+            if (is_object($record)) {
+                return serialize($record);
```
Comment (boajer) [inline models/classes/evnetHandler/ResultTransmissionEventHandler/AsynchronousResultTransmissionEventHandler.php:69]: Nitpicking - but would it be better to use some custom/ safer serialization approach, proposed one have a theoretical risk of becoming attacked

File: models/classes/evnetHandler/ResultTransmissionEventHandler/ResultTransmissionEventHandler.php [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+            $event->getVariables(),
+            $event->getTransmissionId(),
+            $event->getTestUri()
+        );
+    }
+
+    private function buildTransmitter($deliveryExecutionId): taoQtiCommon_helpers_ResultTransmitter
+    {
+        /** @var DeliveryServerService $deliveryServerService */
+        $deliveryServerService = ServiceManager::getServiceManager()->get(DeliveryServerService::SERVICE_ID);
```
Comment (boajer) [inline models/classes/evnetHandler/ResultTransmissionEventHandler/ResultTransmissionEventHandler.php:63]: Would it be possible to use DI here? 

Even if no, you should access service manager like `$this->getServiceManager()`, which is implemented at `InjectionAwareService`

File: models/classes/tasks/ResultTransmission/AbstractResultTransmissionTask.php [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+        }
+        if (count($validationErrors) > 0) {
+            throw new ResultTransmissionTaskValidationException(sprintf(
+                'Absent required parameters: %s',
+                implode(', ', $validationErrors)
+            ));
+        }
+    }
+
+
```
Comment (boajer) [inline models/classes/tasks/ResultTransmission/AbstractResultTransmissionTask.php:62]:
```suggestion

```

File: models/classes/tasks/ResultTransmission/AbstractResultTransmissionTask.php [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    protected function buildTransmitter($deliveryExecutionId): taoQtiCommon_helpers_ResultTransmitter
```
Comment (boajer) [inline models/classes/tasks/ResultTransmission/AbstractResultTransmissionTask.php:63]:
```suggestion
    protected function buildTransmitter(string $deliveryExecutionId): taoQtiCommon_helpers_ResultTransmitter
```

File: models/classes/event/ResultItemVariablesTransmissionEvent.php (hunk 4 above)
Comment (poyuki) [reply models/classes/event/ResultItemVariablesTransmissionEvent.php:30]: can't understand reason for that, could you explain ?

File: models/classes/evnetHandler/ResultTransmissionEventHandler/AsynchronousResultTransmissionEventHandler.php (hunk 6 above)
Comment (poyuki) [reply models/classes/evnetHandler/ResultTransmissionEventHandler/AsynchronousResultTransmissionEventHandler.php:69]: i think you right, but i hasn't any quick solutions in my pocket, and i think it's not really a blocker

File: migrations/Version202112231326462260_taoQtiTest.php [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+
+namespace oat\taoQtiTest\migrations;
+
+use Doctrine\DBAL\Schema\Schema;
+use oat\oatbox\event\EventManager;
+use oat\oatbox\reporting\Report;
+use oat\tao\scripts\tools\migrations\AbstractMigration;
+use oat\taoQtiTest\models\classes\event\ResultTestVariablesTransmissionEvent;
+use oat\taoQtiTest\models\classes\$event\ResultTransmissionEventHandler\Api\ResultTransmissionEventHandlerInterface;
+use oat\taoQtiTest\models\classes\$event\ResultTransmissionEventHandler\AsynchronousResultTransmissionEventHandler;
```
Comment (wazelin) [inline migrations/Version202112231326462260_taoQtiTest.php:13]: Something went very wrong here. What's up with that `$event` in the import?

Comment (poyuki) [reply migrations/Version202112231326462260_taoQtiTest.php]: unsuccessful mass replace, should be fixed already

File: models/classes/event/ResultItemVariablesTransmissionEvent.php (hunk 4 above)
Comment (wazelin) [reply models/classes/event/ResultItemVariablesTransmissionEvent.php:30]: Can't agree with that suggestion either, @boajer. Absolutely no reason to initialize this property as it's being initialized by the constructor.

File: models/classes/event/ResultTestVariablesTransmissionEvent.php (hunk 5 above)
Comment (wazelin) [reply models/classes/event/ResultTestVariablesTransmissionEvent.php:32]: https://github.com/oat-sa/extension-tao-testqti/pull/2156/files#r775028676

File: models/classes/tasks/ResultTransmission/Exception/ResultTransmissionTaskValidationException.php [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2021 (original work) Open Assessment Technologies SA;
+ */
+
+declare(strict_types=1);
+
+namespace oat\taoQtiTest\models\classes\tasks\ResultTransmission\Exception;
+
+class ResultTransmissionTaskValidationException extends \Exception
```
Comment (wazelin) [inline models/classes/tasks/ResultTransmission/Exception/ResultTransmissionTaskValidationException.php:25]: I'd say this kind of an exception should either be extending a `DomainException` or a `RuntimeException`.
I think it should extend the first one, as such an exception means 2 things:
1. The value was correctly typed, but the contents were corrupted
2. Some piece of code was likely to be responsible for the corrupted data, and has to be fixed, which is a definition of a `LogicException`

File: helpers/class.TestSession.php [hunk 12]
```diff
@@ -150,26 +146,6 @@ protected function getResultServer()
...
-    {
-        $this->resultTransmitter = $resultTransmitter;
-    }
-
-    /**
-     * Get the ResultTransmitter object to be used to transmit test results.
-     *
-     * @return taoQtiCommon_helpers_ResultTransmitter
-     */
-    protected function getResultTransmitter()
```
Comment (wazelin) [inline helpers/class.TestSession.php:168]: Removing these two methods is technically a breaking change.
Although, I don't see any other extension overriding or using them, I strongly suggest bumping a major version of the extension, as well as explicitly mentioning what exactly was a breaking change in the future release notes.

File: models/classes/tasks/ResultTransmission/Exception/ResultTransmissionTaskValidationException.php (hunk 11 above)
Comment (poyuki) [reply models/classes/tasks/ResultTransmission/Exception/ResultTransmissionTaskValidationException.php:25]: good point

**Review by wazelin — CHANGES_REQUESTED** [review]
The PR looks really good, leaving just a few notes.
I'm sure you'll finalize it and make available for the QA verification early next week ;)

gj!

File: helpers/class.TestSession.php (hunk 12 above)
Comment (poyuki) [reply helpers/class.TestSession.php:168]: my bad, miss it. will add breaking changes

**Review by wazelin — APPROVED** [review]
Please note I haven't had a chance to check the feature locally myself, but this is something that either would work, or break completely. So, I count on our QA engineers to catch any error. Please merge the PR only _after_ they have confirmed all's well on a test environment.

- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [x] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful
- [x] Pull request's target is not `master`

**Review by boajer — APPROVED** [review]

**Review by yohanlaborda — APPROVED** [review]
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [X] New code is respecting code style rules
- [X] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [X] Acceptance criteria are respected
- [X] Pull request title and description are meaningful
- [X] Pull request's target is not master
- [X] Commits are following conventional commits
- [ ] Changelog is updated according to changes (if applicable)
- [X] Documentation is updated according to changes (if applicable)

**Pull request merged**